platform/darwin/src/NSComparisonPredicate+MGLAdditions.mm

```diff
@@ -330,9 +148,48 @@ - (id)mgl_jsonExpressionObject {
                         format:@"NSPredicateOperatorType:%lu is not supported.", (unsigned long)self.predicateOperatorType];
     }
     if (op) {
-        return @[op, self.leftExpression.mgl_jsonExpressionObject, self.rightExpression.mgl_jsonExpressionObject];
+        id leftExpression = self.leftExpression.mgl_jsonExpressionObject;
+        id rightExpression = self.rightExpression.mgl_jsonExpressionObject;
+        
+        switch (self.predicateOperatorType) {
+            case NSLessThanPredicateOperatorType:
+            case NSLessThanOrEqualToPredicateOperatorType:
+            case NSGreaterThanPredicateOperatorType:
+            case NSGreaterThanOrEqualToPredicateOperatorType: {
+                leftExpression = [self mgl_expressionWithExpressionType:self.leftExpression fallbackExpression:self.rightExpression];
+                rightExpression = [self mgl_expressionWithExpressionType:self.rightExpression fallbackExpression:self.leftExpression];
+                break;
+            }
+ 
+            default:
+                break;
+        }
+        
+        return @[op, leftExpression, rightExpression];
     }
     return nil;
 }
 
+- (id)mgl_expressionWithExpressionType:(NSExpression *)expressionElement fallbackExpression:(NSExpression *)fallbackExpression {
+    NSExpression *expression = expressionElement.expressionType == NSConstantValueExpressionType ? expressionElement : fallbackExpression;
+    NSString *type;
+    if (expression.expressionType == NSConstantValueExpressionType && [expression.constantValue isKindOfClass:NSNumber.class]) {
+        NSNumber *number = (NSNumber *)expression.constantValue;
+        if ((strcmp([number objCType], @encode(char)) == 0) ||
+            (strcmp([number objCType], @encode(BOOL)) == 0)) {
+            type = @"boolean";
+        } else {
+            type = @"number";
+        }
+    } else if (expression.expressionType == NSConstantValueExpressionType && [expression.constantValue isKindOfClass:NSString.class]) {
+        type = @"string";
+    }
+    
+    if (type) {
+        return @[type, expressionElement.mgl_jsonExpressionObject];
```

Comment:
It shouldn’t be the job of the platform-specific code to insert type assertions based on the data type of the constant value.

Reply:
Unfortunately, the code in this PR is only capable of inserting type assertions for the most basic expressions, but not for something more complex like `CAST(ele, 'NSNumber') * 0.3048 > 1000`. To handle the whole gamut of expressions that could require type assertions, we’d have to implement static type analysis in Objective-C.

Since it’s probably too late for mapbox/mapbox-gl-js#6459 to make it into the v4.0.0 release, we’ll have to settle for expecting the developer to include type assertions (or, more likely, `CAST()`) in their predicates. As long as the type checking in this PR isn’t incorrect, I suppose we can keep it, but it would complicate the developer’s mental model of when they’d need to add type assertions themselves.

Reply:
https://github.com/mapbox/mapbox-gl-js/issues/6459 would not take very long to implement, once we have consensus on the design.  Is this something we should try to expedite to get into v4.0.0?

Reply:
@anandthakker if it's possible to get it done for 4.0.0 would be good. Otherwise I think we can live with this for now and in a patch release remove it.

/cc @1ec5

Reply:
Either way, the SDK code should not elide type assertions during conversion.